### PR TITLE
updated change_release to fix regex

### LINKED ISSUE #776: change_release.sh doesn't remove `-dev0`
### Expected behaviour

`maintainer/change_release 0.15.0`

This should remove the `-dev0` ending in all the version strings we have.

### Actual behaviour

Only the numbers are replaced.

I'm not completely understanding the SED regexes that we are using here. The solution would either be to replace the SED regexes or to rewrite the script in python

### PR DESCRIPTION
**Pull request by jdetle**
Changes made in this Pull Request:
* ~~No additional tests added,~~ unsure if necessary. (see below – @orbeckst)
* CHANGELOG updated to reflect fix.
* Fixes #776  

PR Checklist
------------

 - [x] Tests (manual, attached to this PR)
 - [no ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by jbarnoud** [comment]
The change in the regexp looks good to me. I am not familiar with the rest of the script, though.

@jdetle  You should add yourself in the AUTHOR file and in the author list for the last version in the CHANGELOG.

**Comment by kain88-de** [comment]
Can the dev be added again after a release?

**Comment by jdetle** [comment]
@kain88-de I am unclear as to what you mean? `bash maintainer/change_release.sh "0.15.0-dev0` appropriately changes the relevant files if that is what you mean.

**Comment by jdetle** [comment]
One possible issue, in `mdanalysis/maintainer/conda/MDAnalysis/meta.yaml`
in the comments, we have comments saying the "git-tag" is release-0.14.0" should this be changed by the script as well?

**Comment by kain88-de** [comment]
The comment is fine. It is just as a reminder how the commit to be build can be selected as well.

**Comment by orbeckst** [comment]
> Can the dev be added again after a release?

@kain88-de I am not sure I understand your question. Normally, this script is being run right before release (and perhaps @richardjgowers is only running it on master nowadays) and the changes are committed.

Do you mean "Can one use the script to bump the release to the next development release afterwards?", i.e. 

    change_release.sh  0.15.0-dev

The answer *should* be "yes" (at least I'd like to think that this worked in the past).

**Comment by kain88-de** [comment]
> Do you mean "Can one use the script to bump the release to the next development release afterwards?", i.e.

yes thats what I meant. I will have some time tomorrow to check the changes. Feel free to do it yourself if you have time before me.

**Comment by jdetle** [comment]
@orbeckst, if "0.15.0-dev" is a valid name this regex won't work, will need to add two asterisks. Is this the case?

**Comment by richardjgowers** [comment]
I think what's needed is running the script can take the version variable in all concerned files though a sequence like
 - 0.15.0
 - 0.15.1-dev0
 - 0.15.1
 - 0.16.0-dev0
 - 0.16.0
 - etc

So the -dev0 has to be sometimes matched and removed

**Comment by jdetle** [comment]
@richardjgowers that certainly is working, I am going to go ahead and update the PR to commit the possibility of "0.15.1-dev" being a valid name. The appending an asterisk to -dev*[0-9] in the regexes starting at line 76 adds for the possibility of there being no number following dev.

**Comment by orbeckst** [comment]
"0.15.0-dev0" is a release tag that we are using for development code. In the past we have also had tags for release candidates such as "0.15.0-rc1" (and we also used "0.15.0-dev"). If your question is *"is '0.15.0-dev' illegal while '0.15.0-dev0' is legal"* then the answer is that the regex should be tolerant and deal with either.

Our [release numbering](https://github.com/MDAnalysis/mdanalysis/wiki/PreparingReleases#Release_policy_and_release_numbering) is supposed to be compatible with [semantic versioning](https://github.com/MDAnalysis/mdanalysis/wiki/SemanticVersioning).

File: maintainer/change_release.sh [hunk 1]
```diff
@@ -73,13 +73,13 @@ echo "Using sed = $SED"
...
 echo "Setting RELEASE/__version__ in MDAnalysis to $RELEASE"
 
-git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?' $FILES  \
-   | xargs -I FILE $SED '/RELEASE/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]|-rc*[0-9])?' $FILES  \
+   | xargs -I FILE $SED '/RELEASE/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]|-rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
 
-git grep -E -l '__version__ =.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?'  $FILES \
-   | xargs -I FILE $SED '/__version__/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l '__version__=.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]|-rc*[0-9])?'  $FILES \
```
Comment (orbeckst) [inline maintainer/change_release.sh]: The `*` is in the wrong place. Should be
```
(-dev[0-9]*|-rc[0-9]*)
```
so that you can have 0 to many numbers; right now you match `de0`, `devvvvvvvvvvvvvvvvv3`, etc.

(The rc one was apparently an old bug.)

File: maintainer/change_release.sh [hunk 2]
```diff
@@ -73,13 +73,13 @@ echo "Using sed = $SED"
...
+   | xargs -I FILE $SED '/__version__/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]|-rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
```
Comment (orbeckst) [inline maintainer/change_release.sh]: wrong `*`, see above

File: maintainer/change_release.sh [hunk 3]
```diff
@@ -73,13 +73,13 @@ echo "Using sed = $SED"
...
 
-git grep -E -l 'version:.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?'  $FILES \
-   | xargs -I FILE $SED '/version:/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l 'version:.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]|-rc*[0-9])?'  $FILES \
```
Comment (orbeckst) [inline maintainer/change_release.sh]: wrong `*`

File: maintainer/change_release.sh [hunk 4]
```diff
@@ -73,13 +73,13 @@ echo "Using sed = $SED"
...
+   | xargs -I FILE $SED '/version:/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]|-rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
```
Comment (orbeckst) [inline maintainer/change_release.sh]: see above, please fix both for dev and for rc

File: maintainer/change_release.sh [hunk 5]
```diff
@@ -73,13 +73,13 @@ echo "Using sed = $SED"
 
 echo "Setting RELEASE/__version__ in MDAnalysis to $RELEASE"
 
-git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?' $FILES  \
-   | xargs -I FILE $SED '/RELEASE/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]|-rc*[0-9])?' $FILES  \
```
Comment (orbeckst) [inline maintainer/change_release.sh]: wrong position of `*` for dev and rc pattern.

File: maintainer/change_release.sh [hunk 6]
```diff
@@ -73,13 +73,14 @@ echo "Using sed = $SED"
...
 
-git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?' $FILES  \
-   | xargs -I FILE $SED '/RELEASE/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?' $FILES  \
+   | xargs -I FILE $SED
+   '/RELEASE/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?/'${RELEASE}'/' -i.bak FILE
 
-git grep -E -l '__version__ =.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?'  $FILES \
-   | xargs -I FILE $SED '/__version__/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l '__version__=.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?'  $FILES \
```
Comment (orbeckst) [inline maintainer/change_release.sh]: `dev*` is not correct, should be `dev`

You matched, for instance, `devvvvvvvvvvvvvv12`

Same for `-rc*[0-9]*` should be `-rc[0-9]*`.

File: maintainer/change_release.sh [hunk 7]
```diff
@@ -73,13 +73,14 @@ echo "Using sed = $SED"
...
+   | xargs -I FILE $SED '/__version__/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?/'${RELEASE}'/' -i.bak FILE
```
Comment (orbeckst) [inline maintainer/change_release.sh]: wrong `*` placement, see above

File: maintainer/change_release.sh [hunk 8]
```diff
@@ -73,13 +73,14 @@ echo "Using sed = $SED"
...
 
-git grep -E -l 'version:.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?'  $FILES \
-   | xargs -I FILE $SED '/version:/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(dev|rc*[0-9])?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l 'version:.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?'  $FILES \
```
Comment (orbeckst) [inline maintainer/change_release.sh]: wrong `*` placement, see above

File: maintainer/change_release.sh [hunk 9]
```diff
@@ -73,13 +73,14 @@ echo "Using sed = $SED"
...
+   | xargs -I FILE $SED '/version:/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?/'${RELEASE}'/' -i.bak FILE
```
Comment (orbeckst) [inline maintainer/change_release.sh]: wrong `*` placement, see above

**Comment by jdetle** [comment]
@orbeckst yikes, will fix.

**Comment by orbeckst** [comment]
@jdetle , we can't add tests for this but can you please report on how you tested locally that the script works, e.g. the commands that you ran and grep the version lines from the important lines after each invocation of `change_release.sh`?

File: maintainer/change_release.sh (hunk 6 above)
Comment (orbeckst) [inline maintainer/change_release.sh]: P.S.: The `rc*` was there before... so you are also fixing an old bug :-)

**Comment by orbeckst** [comment]
I edited your PR text to include tests as specified above. 

Btw, if you add the line
```
[skip ci]
```
to your commit message, it won't run on Travis (which we don't need for these changes).

**Comment by jdetle** [comment]
[tests.txt](https://github.com/MDAnalysis/mdanalysis/files/180524/tests.txt)
First five lines show successful result after changing to '0.14.1-dev0' 
Second five lines show successful result after changing again to '0.15.0-dev'

File: maintainer/change_release.sh [hunk 10]
```diff
@@ -73,14 +73,13 @@ echo "Using sed = $SED"
...
 
-git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?' $FILES  \
-   | xargs -I FILE $SED
-   '/RELEASE/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l 'RELEASE.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev[0-9]*|-rc[0-9]*)?' $FILES  \
+   | xargs -I FILE $SED '/RELEASE/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev[0-9]*|-rc*[0-9]*)?/'${RELEASE}'/' -i.bak FILE
 
-git grep -E -l '__version__=.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?'  $FILES \
-   | xargs -I FILE $SED '/__version__/s/[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev*[0-9]*|-rc*[0-9]*)?/'${RELEASE}'/' -i.bak FILE
+git grep -E -l '__version__.=.*[0-9]+\.[0-9]+\.[0-9]+(\.[0-9]+)?(-dev[0-9]*|-rc[0-9]*)?'  $FILES \
```
Comment (orbeckst) [inline maintainer/change_release.sh]: Good catch, however, better use
```
__version__ *=
```
to allow any amount of space... maybe `\s*` works but I can't remember if regex has a special "white space" pattern.

**Comment by jdetle** [comment]
[tests.txt](https://github.com/MDAnalysis/mdanalysis/files/180536/tests.txt)

more proof of execution

**Comment by orbeckst** [comment]
Looking good: Two things before merging:
* Please squash *all* commits into a single commit so that the history remains readable. (Have you used `git rebase` before?)
* As a final test (after the squash) please run the following chain of changes and show the results:
 * 0.15.0
 * 0.15.1-dev0
 * 0.15.1
 * 0.16.0-dev0
 * 0.16.0-rc1
 * 0.16.0-rc2
 * 0.16.0

Assuming that this all looks good I will merge.

**Comment by jdetle** [comment]
@orbeckst as you can see, I may have been overzealous in my commit squashing. This is my first pull request and first time rebasing commits. Commit fc0d3ce seems to have squashed successfully, but please tell me if I did something wrong.  I recognize it is not your job to teach me how to use github, https://git-scm.com/docs/git-rebase I used this as source material and followed it to the letter. I just want to make sure I didn't do anything wrong, thanks in advance. 
[tests.txt](https://github.com/MDAnalysis/mdanalysis/files/180582/tests.txt)
Attached are the 7 test cases you wanted me to run as a final check.

**Comment by orbeckst** [comment]
There's still one `rc*`. 

I'm not sure if the squashing really worked because when I look at the commits then I see lots of other commits showing up. On my mobile phone it's a bit hard to parse what's going on. Rebasing is not easy so I applause you for learning about it and trying it  (I use `git rebase -i` for squashing). 

Just add the small fix and one of us will either provide more detailed instructions or do the squash.

**Comment by jdetle** [comment]
@orbeckst thanks so much. It looks as though if I use fixup instead of squash in the interactive rebase environment it would delete the old commits from the log which is what I believe we are seeing. I'll go ahead and try that and see if that fixes things.

**Comment by jdetle** [comment]
I used git rebase -i too, but I think instead of squash I wanted to fix

**Comment by kain88-de** [comment]
@jdetle that looks like what you want to do. If you aren't sure you can create a backup branch `git branch change-release-backup`

**Comment by jdetle** [comment]
@orbeckst  @jbarno  @kain88-de Okay, I tried and ended up just creating two more commits. I can imagine this is frustrating for you all. On the bright side, I think I understand what I am doing wrong. After I rebase commits, `git log` shows only one commit made by me, which is what I want. When trying to use `git push origin`  the push command states I am behind in time with respect to the develop branch and asks me to pull from the develop branch. This has been the mistake on my part, because when I pull from the develop branch I add all the old commits from the log that I was trying to rebase into one commit. What (I think) I need to do  is use `git push -f origin` to "go back in time" with my commit. Does this sound familiar/ like the right course of action? I appreciate your help.

**Comment by jbarnoud** [comment]
I am not sure of what goes wrong with your rebase. First of all, I would work on a copy of the branch:

```bash
git checkout -b issue-776-change_release
```

This should create a copy of your `develop` branch. Then I would update your develop branch to match the current status of MDAnalysis `develop` branch:

```bash
# Get easy access to the main MDAnalysis repo
# This creates a `remote` named `upstream`
git remote add upstream git@github.com:MDAnalysis/mdanalysis.git

# Get the up to date status of upstream
git fetch upstream develop
```

Now that you have a copy of the main MDAnalysis `develop` branch, you can rebase against it:

```bash
git rebase -i upstream/develop
```

You should get one commit of yours as the most recent commit. Push your `issue-776-change_release` branch and try to make a fake pull request (fake in the sense that you do not have to submit it). Inspect the diff to see if it looks as you want. If so you an replace your `develop` branch by your `issue-776-change_release` one:

```bash
# Delete your develop branch
git branch -D develop
# Create a new develop branch that is a copy of issue-776-change_release
git checkout -b develop
```

Move to your new `develop` branch and make sure it looks as it should:

```bash
git checkout develop
git status
git log
```

If everything is fine, you can push the `develop` branch to github to update the PR. You will have to force the push, which can lead to data loss if anything goes wrong. So, at that point, you need to be sure your `develop` branch is how you want it to be.

```bash
git push origin develop --force
```

Hopefully everything is beautiful, and the PR is perfect.

Next time, work directly in a separate branch, so you can keep your `develop` branch in sync with MDAnalysis one.

**Comment by jdetle** [comment]
Success! I am in your debt @jbarnoud

**Comment by jbarnoud** [comment]
Glad it worked :smile:. Could you edit the commit message for something less messy and more descriptive? We are not always as meticulous as we should be with commit messages, possibly because we rely a lot on the PR to keep track of the history. But a good commit message is important to keep track of what is done and of the motivations behind a change. At this point, you should be able to edit the commit message without going through the pain of a rebase by amending the last commit:

```bash
git commit --amend
# This should open your text editor with the current commit message.
# Replace the message by a neat one explaining what you did and why.
# Do not forget to reference the issue and the PR.
git push --force
```

**Comment by jdetle** [comment]
Of course, I am glad I am being forced to have good habits from the beginning.

**Comment by jbarnoud** [comment]
It looks good to me. Since @orbeckst is assigned, I let him have the last look before merge.

**Comment by kain88-de** [comment]
@jdetle rebasing is fine and once you get used to it it isn't any different from using amend. In fact when I'm developing a feature I locally do tens of small commits. (Trying to restrict myself to one file per commit). In the end I reorder and combine commits with `rebase` into logical units. One of the nice things in git that I can to that.

To learn the ins and outs of git and how it work I can heartly recommend [this](http://rypress.com/tutorials/git/index) tutorial.

**Comment by orbeckst** [comment]
I'm busy for the day, please anyone else grab it. Thanks!

--
Oliver Beckstein
email: orbeckst@gmail.com


Am Mar 19, 2016 um 13:03 schrieb Jonathan Barnoud <notifications@github.com>:

**Comment by jbarnoud** [comment]
@jdetle Congratulations for your first merged PR. I hope we will see plenty of other ones.
